File: app/src/main/java/com/squareoverlay/SquareOverlayView.java
package com.squareoverlay;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.view.WindowManager;

public class SquareOverlayView extends View {

    private static final String TAG = "SquareOverlay";
    
    private Paint squarePaint;
    private Paint textPaint;
    private Paint textBackgroundPaint;
    
    private WindowManager windowManager;
    private WindowManager.LayoutParams windowParams;
    
    private int screenWidth;
    private int screenHeight;
    
    private float squareX;
    private float squareY;
    private float squareSize;
    
    private float absoluteScreenX;
    private float absoluteScreenY;
    
    private float initialTouchX;
    private float initialTouchY;
    private int initialWindowX;
    private int initialWindowY;
    private boolean isDragging = false;
    
    private static final int CORNER_SIZE = 160;
    private boolean isResizing = false;

    public interface ScreenshotCallback {
        void onScreenshotRequested(float xPercent, float yPercent, float widthPercent, float heightPercent, Runnable onHidden);
    }

    private ScreenshotCallback screenshotCallback;

    public SquareOverlayView(Context context, int screenWidth, int screenHeight, float initialSquareSize) {
        super(context);
        
        this.screenWidth = screenWidth;
        this.screenHeight = screenHeight;
        
        squareSize = initialSquareSize;
        squareX = 0;
        squareY = 0;

        absoluteScreenX = (screenWidth - squareSize) / 2;
        absoluteScreenY = (screenHeight - squareSize) / 2;
        
        squarePaint = new Paint();
        squarePaint.setColor(Color.RED);
        squarePaint.setStyle(Paint.Style.STROKE);
        squarePaint.setStrokeWidth(5);
        squarePaint.setAntiAlias(true);
        
        textPaint = new Paint();
        textPaint.setColor(Color.WHITE);
        textPaint.setTextSize(40);
        textPaint.setAntiAlias(true);
        
        textBackgroundPaint = new Paint();
        textBackgroundPaint.setColor(Color.argb(200, 0, 0, 0));
        textBackgroundPaint.setStyle(Paint.Style.FILL);

        Log.d(TAG, "View created with square size: " + (int)squareSize);
    }

    public void setScreenshotCallback(ScreenshotCallback callback) {
        this.screenshotCallback = callback;
    }

    public void setWindowManager(WindowManager wm, WindowManager.LayoutParams params) {
        this.windowManager = wm;
        this.windowParams = params;
    }

    public void triggerScreenshot() {
        if (screenshotCallback != null) {
            float xPercent = (absoluteScreenX / screenWidth) * 100;
            float yPercent = (absoluteScreenY / screenHeight) * 100;
            float widthPercent = (squareSize / screenWidth) * 100;
            float heightPercent = (squareSize / screenHeight) * 100;

            screenshotCallback.onScreenshotRequested(xPercent, yPercent, widthPercent, heightPercent, () -> {
                // This runs after screenshot is hidden
            });
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        // View size matches square size exactly
        int width = (int)squareSize;
        int height = (int)squareSize;
        setMeasuredDimension(width, height);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        
        canvas.drawRect(squareX, squareY, squareX + squareSize, squareY + squareSize, squarePaint);
        
        // Draw resize handle (reuse paint object)
        squarePaint.setStyle(Paint.Style.FILL);
        squarePaint.setColor(Color.BLUE);
        canvas.drawCircle(squareX + squareSize, squareY + squareSize, 55, squarePaint);
        squarePaint.setColor(Color.RED);
        squarePaint.setStyle(Paint.Style.STROKE);
        
        if (windowParams != null) {
            absoluteScreenX = windowParams.x + squareX;
            absoluteScreenY = windowParams.y + squareY;
        }
        
        float xPercent = (absoluteScreenX / screenWidth) * 100;
        float yPercent = (absoluteScreenY / screenHeight) * 100;
        float sizePercent = (squareSize / screenWidth) * 100;

        String line1 = String.format("Position: %d, %d", (int)absoluteScreenX, (int)absoluteScreenY);
        String line2 = String.format("Size: %d x %d", (int)squareSize, (int)squareSize);
        String line3 = String.format("Crop: %.1f%%, %.1f%%, %.1f%%", xPercent, yPercent, sizePercent);

        float textX = squareX + 20;
        float textY = squareY + 60;
        float lineHeight = 50;

        float textWidth = 380;
        float textHeight = 3 * lineHeight + 20;
        canvas.drawRect(textX - 10, textY - 45, textX + textWidth, textY + textHeight - 45, textBackgroundPaint);

        canvas.drawText(line1, textX, textY, textPaint);
        canvas.drawText(line2, textX, textY + lineHeight, textPaint);
        canvas.drawText(line3, textX, textY + lineHeight * 2, textPaint);
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        if (windowManager == null || windowParams == null) {
            return false;
        }
        
        float touchX = event.getX();
        float touchY = event.getY();
        
        switch (event.getAction()) {
            case MotionEvent.ACTION_DOWN:
                float cornerX = squareX + squareSize;
                float cornerY = squareY + squareSize;
                float distanceToCorner = (float) Math.sqrt(
                    Math.pow(touchX - cornerX, 2) + Math.pow(touchY - cornerY, 2)
                );

                if (distanceToCorner < CORNER_SIZE) {
                    isResizing = true;
                    isDragging = false;
                    Log.d(TAG, "Started resizing");
                    return true;
                } else if (touchX >= squareX && touchX <= squareX + squareSize &&
                           touchY >= squareY && touchY <= squareY + squareSize) {
                    isDragging = true;
                    isResizing = false;
                    initialTouchX = event.getRawX();
                    initialTouchY = event.getRawY();
                    initialWindowX = windowParams.x;
                    initialWindowY = windowParams.y;
                    Log.d(TAG, "Started dragging");
                    return true;
                }
                return false;
                
            case MotionEvent.ACTION_MOVE:
                if (isResizing) {
                    float newSize = Math.max(200, touchX - squareX);
                    newSize = Math.min(newSize, Math.min(screenWidth * 0.9f, screenHeight * 0.9f));

                    if (newSize != squareSize) {
                        squareSize = newSize;

                        // Update window size to match new square size
                        windowParams.width = (int)squareSize;
                        windowParams.height = (int)squareSize;

                        windowManager.updateViewLayout(this, windowParams);

                        Log.d(TAG, "Resized: size=" + (int)squareSize);
                    }
                    return true;
                } else if (isDragging) {
                    float deltaX = event.getRawX() - initialTouchX;
                    float deltaY = event.getRawY() - initialTouchY;

                    windowParams.x = initialWindowX + (int)deltaX;
                    windowParams.y = initialWindowY + (int)deltaY;

                    // Keep square on screen - window position = square position
                    int minX = 0;
                    int minY = 0;
                    int maxX = screenWidth - (int)squareSize;
                    int maxY = screenHeight - (int)squareSize;

                    windowParams.x = Math.max(minX, Math.min(windowParams.x, maxX));
                    windowParams.y = Math.max(minY, Math.min(windowParams.y, maxY));

                    windowManager.updateViewLayout(this, windowParams);
                    invalidate();
                    return true;
                }
                return false;
                
            case MotionEvent.ACTION_UP:
            case MotionEvent.ACTION_CANCEL:
                if (isDragging || isResizing) {
                    Log.d(TAG, "Final: x=" + (int)absoluteScreenX + ", y=" + (int)absoluteScreenY + ", size=" + (int)squareSize);
                }
                isDragging = false;
                isResizing = false;
                return true;
        }
        
        return false;
    }
}

File: app/src/main/java/com/squareoverlay/SquareOverlayView.kt
package com.squareoverlay

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.view.MotionEvent
import android.view.View
import android.view.WindowManager
import kotlin.math.max
import kotlin.math.min
import kotlin.math.pow
import kotlin.math.sqrt

class SquareOverlayView(
    context: Context?,
    private val screenWidth: Int,
    private val screenHeight: Int,
    private var squareSize: Float
) : View(context) {
    private val squarePaint: Paint
    private val textPaint: Paint
    private val textBackgroundPaint: Paint

    private var windowManager: WindowManager? = null
    private var windowParams: WindowManager.LayoutParams? = null

    private val squareX = 0f
    private val squareY = 0f

    private var absoluteScreenX: Float
    private var absoluteScreenY: Float

    private var initialTouchX = 0f
    private var initialTouchY = 0f
    private var initialWindowX = 0
    private var initialWindowY = 0
    private var isDragging = false

    private var isResizing = false

    fun interface ScreenshotCallback {
        fun onScreenshotRequested(
            xPercent: Float,
            yPercent: Float,
            widthPercent: Float,
            heightPercent: Float,
            onHidden: Runnable?
        )
    }

    private var screenshotCallback: ScreenshotCallback? = null

    init {
        absoluteScreenX = (screenWidth - squareSize) / 2
        absoluteScreenY = (screenHeight - squareSize) / 2

        squarePaint = Paint().apply {
            color = Color.RED
            style = Paint.Style.STROKE
            strokeWidth = 5f
            isAntiAlias = true
        }

        textPaint = Paint().apply {
            color = Color.WHITE
            textSize = 40f
            isAntiAlias = true
        }

        textBackgroundPaint = Paint().apply {
            color = Color.argb(200, 0, 0, 0)
            style = Paint.Style.FILL
        }
    }

    fun setScreenshotCallback(callback: ScreenshotCallback?) {
        this.screenshotCallback = callback
    }

    fun setWindowManager(wm: WindowManager?, params: WindowManager.LayoutParams?) {
        this.windowManager = wm
        this.windowParams = params
    }

    fun triggerScreenshot() {
        screenshotCallback?.onScreenshotRequested(
            (absoluteScreenX / screenWidth) * 100,
            (absoluteScreenY / screenHeight) * 100,
            (squareSize / screenWidth) * 100,
            (squareSize / screenHeight) * 100,
            Runnable {}
        )
    }

    fun setSquareSize(size: Float) {
        this.squareSize = size
        requestLayout()
        invalidate()
    }

    val squareXPercent: Float
        get() {
            windowParams?.let {
                absoluteScreenX = it.x + squareX
            }
            return (absoluteScreenX / screenWidth) * 100
        }

    val squareYPercent: Float
        get() {
            windowParams?.let {
                absoluteScreenY = it.y + squareY
            }
            return (absoluteScreenY / screenHeight) * 100
        }

    val squareWidthPercent: Float
        get() = (squareSize / screenWidth) * 100

    val squareHeightPercent: Float
        get() = (squareSize / screenHeight) * 100

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        setMeasuredDimension(squareSize.toInt(), squareSize.toInt())
    }

    override fun onDraw(canvas: Canvas) {
        super.onDraw(canvas)

        canvas.drawRect(squareX, squareY, squareX + squareSize, squareY + squareSize, squarePaint)

        squarePaint.apply {
            style = Paint.Style.FILL
            color = Color.BLUE
        }
        canvas.drawCircle(squareX + squareSize, squareY + squareSize, 55f, squarePaint)
        squarePaint.apply {
            color = Color.RED
            style = Paint.Style.STROKE
        }

        windowParams?.let {
            absoluteScreenX = it.x + squareX
            absoluteScreenY = it.y + squareY
        }

        val xPercent = (absoluteScreenX / screenWidth) * 100
        val yPercent = (absoluteScreenY / screenHeight) * 100
        val sizePercent = (squareSize / screenWidth) * 100

        val line1 = "Position: ${absoluteScreenX.toInt()}, ${absoluteScreenY.toInt()}"
        val line2 = "Size: ${squareSize.toInt()} x ${squareSize.toInt()}"
        val line3 = String.format("Crop: %.1f%%, %.1f%%, %.1f%%", xPercent, yPercent, sizePercent)

        val textX = squareX + 20
        val textY = squareY + 60
        val lineHeight = 50f

        val textWidth = 380f
        val textHeight = 3 * lineHeight + 20
        canvas.drawRect(
            textX - 10,
            textY - 45,
            textX + textWidth,
            textY + textHeight - 45,
            textBackgroundPaint
        )

        canvas.drawText(line1, textX, textY, textPaint)
        canvas.drawText(line2, textX, textY + lineHeight, textPaint)
        canvas.drawText(line3, textX, textY + lineHeight * 2, textPaint)
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        val wm = windowManager ?: return false
        val params = windowParams ?: return false

        val touchX = event.x
        val touchY = event.y

        when (event.action) {
            MotionEvent.ACTION_DOWN -> {
                val cornerX = squareX + squareSize
                val cornerY = squareY + squareSize
                val distanceToCorner = sqrt(
                    (touchX - cornerX).pow(2) + (touchY - cornerY).pow(2)
                )

                return when {
                    distanceToCorner < CORNER_SIZE -> {
                        isResizing = true
                        isDragging = false
                        true
                    }
                    touchX >= squareX && touchX <= squareX + squareSize
                            && touchY >= squareY && touchY <= squareY + squareSize -> {
                        isDragging = true
                        isResizing = false
                        initialTouchX = event.rawX
                        initialTouchY = event.rawY
                        initialWindowX = params.x
                        initialWindowY = params.y
                        true
                    }
                    else -> false
                }
            }

            MotionEvent.ACTION_MOVE -> {
                when {
                    isResizing -> {
                        var newSize = max(200f, touchX - squareX)
                        newSize = min(newSize, min(screenWidth * 0.9f, screenHeight * 0.9f))

                        if (newSize != squareSize) {
                            squareSize = newSize
                            params.width = squareSize.toInt()
                            params.height = squareSize.toInt()
                            wm.updateViewLayout(this, params)
                        }
                        true
                    }
                    isDragging -> {
                        val deltaX = event.rawX - initialTouchX
                        val deltaY = event.rawY - initialTouchY

                        params.x = (initialWindowX + deltaX.toInt()).coerceIn(
                            0,
                            screenWidth - squareSize.toInt()
                        )
                        params.y = (initialWindowY + deltaY.toInt()).coerceIn(
                            0,
                            screenHeight - squareSize.toInt()
                        )

                        wm.updateViewLayout(this, params)
                        invalidate()
                        true
                    }
                    else -> false
                }
            }

            MotionEvent.ACTION_UP, MotionEvent.ACTION_CANCEL -> {
                isDragging = false
                isResizing = false
                true
            }

            else -> false
        }

        return false
    }

    companion object {
        private const val CORNER_SIZE = 160
    }
}
